keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1086 @@
+resources. In the interest of supporting mutating admission controllers, policies will only
+be enforced on actual pods.
+
+Templated pod resources include:
+
+- v1 ReplicationController
+- v1 PodTemplate
+- apps/v1 ReplicaSet
+- apps/v1 Deployment
+- apps/v1 StatefulSet
+- apps/v1 DaemonSet
+- batch/v1 CronJob
+- batch/v1 Job
+
+PodTemplate warnings & audit will only be applied to built-in types. CRDs that wish to take
+advantage of this functionality can use an object reference to a v1/PodTemplate resource rather than
+inlining a PodTemplate. We will publish a guide (documentation and/or examples) that demonstrate
+this pattern. Alternatively, the functionality can be implemented in a 3rd party admission plugin
+leveraging the library implementation.
+
+### Namespace policy update warnings
+
+When an `enforce` policy (or version) label is added or changed, the admission plugin will test each pod
+in the namespace against the new policy. Violations are returned to the user as warnings. These
+checks have a timeout of XX seconds and a limit of YY pods, and will return a warning in the event
+that not every pod was checked. User exemptions are ignored by these checks, but runtime class
+exemptions still apply. Namespace exemptions are also ignored, but an additional warning will be
+returned when updating the policy on an exempt namespace. These checks only consider actual Pod
+resources, not [templated pods].
+
+These checks are also performed when making a dry-run request, which can be an effective way of
+checking for breakages before updating a policy, for example:
+
+```
+kubectl label --dry-run=server --overwrite ns --all <prefix>/enforce=baseline
+```
+
+<<[UNRESOLVED]>>
+
+_Non-blocking: can be decided on the implementing PR_
+
+- What should the timeout be for pod update warnings?
+  - Total is a parameter on the context (query parameter for webhooks). Cap should be
+    `min(timeout_param, hard_cap)`, where the `hard_cap` is a small number of seconds.
+  - Expect evaluation to be fast, so even 3k pods should come in well under the timeout.
+- What should the pod limit be set to?
+  - 3,000 is the
+    [documented](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md)
+    scalability limit for per-namespace pod count.
+  - Warnings should be aggregated for large namespaces (soft cap number of warnings, hard cap number
+    of evaluations).
+
+<<[/UNRESOLVED]>>
+
+
+### Admission Configuration
+
+A number of options can be statically configured through the [Admission Configuration file][]:
+
+```
+apiVersion: apiserver.config.k8s.io/v1
+kind: AdmissionConfiguration
+plugins:
+- name: PSPReplacement # Placeholder
+  configuration:
+    defaults:  # Defaults applied when a mode label is not set.
+      allow:         <default allow policy level>
+      allow-version: <default allow policy version>
```

Comment:
are these supposed to be `enforce` and `enforce-version`?

Reply:
yes

Reply:
yes, thanks.